arch/i386/mcount-support.c

```diff
@@ -0,0 +1,354 @@
+
+		pr_dbg3("[%d] %s got idx %d, r_offset %lx, real address = %#lx, "
+			"target addr = %p, jump offset = %#lx\n",
+			idx, sym->name, got_idx, r_offset + sym->addr - pd->base_addr + JMP_INSN_SIZE,
+			real_addr, target_addr, jump_offset);
+
+		/* make up the instruction and copy to the trampoline buffer */
+		memcpy(&trampoline[PUSH_IDX_POS], &idx, sizeof(idx));
+		memcpy(&trampoline[JMP_OFS_POS], &jump_offset, sizeof(jump_offset));
+		memcpy(target_addr, trampoline, sizeof(trampoline));
+	}
+
+	pr_dbg2("real address to jump: %#lx\n", plt_addr);
+	memcpy(trampoline_buf + (idx * sizeof(trampoline)),
+	       &plt_addr, sizeof(plt_addr));
+
+	mprotect(trampoline_buf, trampoline_size, PROT_READ | PROT_EXEC);
+	return 0;
+}
+
+unsigned long mcount_arch_plthook_addr(struct plthook_data *pd, int idx)
+{
+	struct sym *sym;
+
+	sym = &pd->dsymtab.sym[idx];
+	return sym->addr + ARCH_PLTHOOK_ADDR_OFFSET;
+}
+
+
+/*
+	For 16-byte stack-alignment, 
+	the main function stores the return address in its stack scope at prologue.
+	When the time comes for the main function to return,
+	1. restore the saved return address from stack.
+	2. After cleaning up the stack.
+	3. Put the return address at the top of the stack and return.
+	4. will be returned.
+
+	080485f8 <main>:
+	80485f8: 8d 4c 24 04           lea    0x4(%esp),%ecx
+	80485fc: 83 e4 f0              and    $0xfffffff0,%esp
+	80485ff: ff 71 fc              pushl  -0x4(%ecx)
+	8048602: 55                    push   %ebp
+	8048603: 89 e5                 mov    %esp,%ebp
+	8048605: 51                    push   %ecx
+	8048606: 83 ec 14              sub    $0x14,%esp
+	8048609: e8 02 fe ff ff        call   8048410 <mcount@plt>
+
+	... ... 
+
+	8048645: 8b 4d fc              mov    -0x4(%ebp),%ecx
+	8048648: c9                    leave
+	8048649: 8d 61 fc              lea    -0x4(%ecx),%esp
+	804864c: c3                    ret
+
+	So, in this case. The return address we want to replace with 
+	mcount_exit is in the stack scope of the main function. 
+	Non a parent located. 
+
+	we search stack for that address. 
+	we will look for it.
+	we will find it, and we will replace it. 
+	GOOD LUCK!
+*/
+unsigned long *mcount_arch_parent_location(struct symtabs *symtabs,
+                                            unsigned long *parent_loc, 
+                                            unsigned long child_ip)
+{
+	if (!main_ret_found) {
+		struct sym *parent_sym, *child_sym;
+		char *pname, *cname;
+
+		const char *find_main[] = {
+			"__libc_start_main",
+			"main"
+		};
+		unsigned long ret_addr;
+		unsigned long search_ret_addr;
+
+		ret_addr = *parent_loc;
+		parent_sym = find_symtabs(symtabs, ret_addr);
+		pname = symbol_getname(parent_sym, ret_addr);
+		child_sym = find_symtabs(symtabs, child_ip);
+		cname = symbol_getname(child_sym, child_ip);
+
+		// Assuming that this happens only in main.			
+		if (!strcmp(find_main[0], pname)) {
+			if (!strcmp(find_main[1], cname)) {
```

Comment:
I prefer bailing out the false condition early and keep the nesting level small.  Like below?

```
if (strcmp(find_main[0], pname) || strcmp(find_main[1], cname))
    goto out;
```

Reply:
it's good, I'll change name of variable about pname and cname for Clarify the meaning. 
pname to parent_name, cname to called_name.
how about it?